datafusion/datasource/src/source.rs

```diff
@@ -175,6 +175,23 @@ impl ExecutionPlan for DataSourceExec {
         self.data_source.statistics()
     }
 
+    fn statistics_by_partition(&self) -> datafusion_common::Result<Vec<Statistics>> {
+        let mut statistics = vec![
+            Statistics::new_unknown(&self.schema());
+            self.properties().partitioning.partition_count()
+        ];
+        if let Some(file_config) =
+            self.data_source.as_any().downcast_ref::<FileScanConfig>()
+        {
+            for (idx, file_group) in file_config.file_groups.iter().enumerate() {
+                if let Some(stat) = file_group.statistics() {
+                    statistics[idx] = stat.clone();
```

Comment:
I am also growing worried about the amount of cloning happening for each Statistics object... they are deep clones at the moment

Reply:
Yes, I have some thoughts.

Reply:
Yes, at least we can use `Arc` now. I've done a pre PR to add Arc for the statistics `FileGroup` https://github.com/apache/datafusion/pull/15564

Reply:
> we can utilize `PlanProperties`? The Statistics will be initiated once and cached like other planning properties

I've had a general look at it and it should work.